### PR TITLE
Update catalogue item to use manufacturer id #102

### LINKED ISSUE #102: Update catalogue item to use manufacturer ID

### PR DESCRIPTION
**Pull request by MatteoGuarnaccia5**
## Description
Changed catalogue items to use manufacturer IDs instead of a manufacturer object. Implemented error checking to ensure manufacturer ID entered in post and patch requests is valid and exists in the database. Updated all unit and e2e tests so that they use manufacturer ID and pass

## Testing instructions
Add a set up instructions describing how the reviewer should test the code

- [ ] Review code
- [ ] Check Actions build
- [ ] Review changes to test coverage

## Agile board tracking
closes #102

### CODE REVIEW

File: test/e2e/test_catalogue_item.py [hunk 1]
```diff
@@ -815,25 +877,23 @@ def test_partial_update_catalogue_item_change_value_for_boolean_property_invalid
 
 def test_partial_update_catalogue_item_change_manufacturer(test_client):
```
Comment (VKTB) [inline test/e2e/test_catalogue_item.py:878]:
```suggestion
def test_partial_update_catalogue_item_change_manufacturer_id(test_client):
```

File: inventory_management_system_api/models/catalogue_item.py [hunk 2]
```diff
@@ -44,9 +27,7 @@ class CatalogueItemIn(BaseModel):
     name: str
     description: str
     properties: List[Property] = []
-    # pylint: disable=fixme
-    # TODO - Change from manufacturer to manufacturer id
-    manufacturer: Manufacturer
+    manufacturer_id: str
```
Comment (VKTB) [inline inventory_management_system_api/models/catalogue_item.py:30]: Its type should be `CustomObjectIdField`. Could you please also update the repo unit tests accordingly?
```suggestion
    manufacturer_id: CustomObjectIdField
```

File: test/unit/services/test_catalogue_item.py
Comment (VKTB) [inline test/unit/services/test_catalogue_item.py:1]: We should also test:
- creating a catalogue item with a non-existent manufacturer ID.
- changing/updating the manufacturer ID of a catalogue ID to an existent as well as a non-existent ID.

File: test/e2e/test_catalogue_item.py [hunk 3]
```diff
@@ -815,25 +877,23 @@ def test_partial_update_catalogue_item_change_value_for_boolean_property_invalid
...
     """
-    Test updating the manufacturer details of a catalogue item.
+    Test updating the manufacturer ID of a catalogue item.
     """
     # pylint: disable=fixme
```
Comment (VKTB) [inline test/e2e/test_catalogue_item.py:882]:
```suggestion
```

File: test/e2e/test_catalogue_item.py [hunk 4]
```diff
@@ -844,20 +904,52 @@ def test_partial_update_catalogue_item_change_manufacturer(test_client):
...
-    assert catalogue_item["manufacturer"] == catalogue_item_patch["manufacturer"]
+    assert catalogue_item["manufacturer_id"] == catalogue_item_patch["manufacturer_id"]
 
 
-def test_partial_update_catalogue_item_change_manufacturer_id_invalid_id():
+def test_partial_update_catalogue_item_change_manufacturer_id_invalid_id(test_client):
     """
     Test changing the manufacturer ID of a catalogue item to an invalid ID.
     """
     # pylint: disable=fixme
```
Comment (VKTB) [inline test/e2e/test_catalogue_item.py:914]:
```suggestion
```

File: test/e2e/test_catalogue_item.py [hunk 5]
```diff
@@ -844,20 +904,52 @@ def test_partial_update_catalogue_item_change_manufacturer(test_client):
...
+
+    assert response.status_code == 422
+    assert response.json()["detail"] == "The specified manufacturer ID does not exist"
+
+
+def test_partial_update_catalogue_item_change_manufacturer_id_nonexistent_id(test_client):
     """
     Test changing the manufacturer ID of a catalogue item to a nonexistent ID.
     """
     # pylint: disable=fixme
```
Comment (VKTB) [inline test/e2e/test_catalogue_item.py:938]:
```suggestion
```

**Review by VKTB — CHANGES_REQUESTED** [review]
Looks good, just a few minor comments. In addition to these, there is also the `test_delete_manufacturer_that_is_a_part_of_catalogue_item` in `e2e\test_manufacturer.py` that needs to be enabled now.

File: inventory_management_system_api/schemas/catalogue_item.py [hunk 6]
```diff
@@ -46,7 +38,7 @@ class CatalogueItemPostRequestSchema(BaseModel):
     )
     # pylint: disable=fixme
     # TODO - Change from manufacturer to manufacturer id
```
[the remark this answers is not in the document]
Comment (VKTB) [reply inventory_management_system_api/schemas/catalogue_item.py]: @MatteoGuarnaccia5 This has not been fully addressed as `# pylint: disable=fixme` is still there.

File: inventory_management_system_api/models/catalogue_item.py (hunk 2 above)
Comment (VKTB) [reply inventory_management_system_api/models/catalogue_item.py]: @MatteoGuarnaccia5 There are some repo unit tests that have not been updated to return the `manufacturer_id` as `CustomObjectId` object. Have a look at the `test_update` test and some of the `test_list_...` tests.

**Review by VKTB — CHANGES_REQUESTED** [review]
Just a few more minor stuff I noticed in your recent changes. Like I highlighted in my previous review, could you please enable the `test_delete_manufacturer_that_is_a_part_of_catalogue_item` in `e2e\test_manufacturer.py`?

File: test/unit/services/test_catalogue_item.py [hunk 7]
```diff
@@ -922,6 +925,101 @@ def test_update_with_nonexistent_catalogue_category_id(
...
+        **catalogue_item_info,
+        "manufacturer_id": str(ObjectId()),
+    }
+    catalogue_item = CatalogueItemOut(**full_catalogue_item_info)
+
+    # Mock `get` to return a catalogue item
+    test_helpers.mock_get(
+        catalogue_item_repository_mock,
+        CatalogueItemOut(
+            manufacturer_id=catalogue_item.manufacturer_id,
```
Comment (VKTB) [inline test/unit/services/test_catalogue_item.py:957]: If updating it to a different `manufacturer_id` then the catalogue item coming from the database should have a different `manufacturer_id` to the one in the updated `catalogue_item`.
```suggestion
            manufacturer_id=str(ObjectId()),
```

**Review by VKTB — DISMISSED** [review]

File: test/e2e/test_catalogue_item.py [hunk 8]
```diff
@@ -975,20 +1315,57 @@ def test_partial_update_catalogue_item_change_manufacturer(test_client):
...
         **catalogue_item_patch,
         "catalogue_category_id": catalogue_category_id,
+        "manufacturer_id": manufacturer_e_id,
     }
 
 
-def test_partial_update_catalogue_item_change_manufacturer_id_invalid_id():
-    """
+def test_partial_update_catalogue_item_change_manufacturer_id_invalid_id(test_client):
+    """/v1/manufacturers
```
Comment (VKTB) [inline test/e2e/test_catalogue_item.py:1323]:
```suggestion
    """
```

File: test/e2e/test_manufacturer.py [hunk 9]
```diff
@@ -399,48 +403,56 @@ def test_delete_with_a_nonexistent_id(test_client):
...
+    "drawing_link": "https://drawing-link.com/",
+    "item_model_number": "abc123",
+    "is_obsolete": False,
+    "properties": [
+        {"name": "Property A", "value": 20},
+        {"name": "Property B", "value": False},
+        {"name": "Property C", "value": "20x15x10"},
+    ],
+    "manufacturer_id": manufacturer_id
+}
```
Comment (VKTB) [inline test/e2e/test_manufacturer.py:451]:
```suggestion
    catalogue_item_post = {
        "name": "Catalogue Item A",
        "catalogue_category_id": catalogue_category_id,
        "description": "This is Catalogue Item A",
        "cost_gbp": 129.99,
        "days_to_replace": 2.0,
        "drawing_link": "https://drawing-link.com/",
        "item_model_number": "abc123",
        "is_obsolete": False,
        "properties": [
            {"name": "Property A", "value": 20},
            {"name": "Property B", "value": False},
            {"name": "Property C", "value": "20x15x10"},
        ],
        "manufacturer_id": manufacturer_id
    }
```

**Review by joelvdavies — APPROVED** [review]
LGTM - To be merged with https://github.com/ral-facilities/inventory-management-system/pull/144

**Review by VKTB — APPROVED** [review]
LGTM. I fixed the formatting and did a minor refactor in the tests.

**Pull request merged**